readme.txt
=== TokenICO ===
Contributors: ethereumicoio, freemius
Tags: ethereum, erc20, ICO, initial coin offering, crypto, cryptocurrency
Requires at least: 3.7
Tested up to: 5.8.1
Stable tag: 2.3.9
Donate link: https://etherscan.io/address/0x476Bb28Bc6D0e9De04dB5E19912C392F9a76535d
License: GPLv2 or later
License URI: https://www.gnu.org/licenses/gpl-2.0.html
Requires PHP: 5.6

Sell your Ethereum ERC20 ICO tokens from your WordPress site.

== Description ==

Ethereum ICO Wordpress plugin can be used to sell your Ethereum ERC20 ICO tokens from your Wordpress site.

> It is the only available WP plugin to sell your Ethereum ERC20 ICO tokens directly from your WordPress site.

> Binance Smart Chain (BSC), Polygon (MATIC) and any other EVM-compatible blockchain is supported in the [PRO version](https://checkout.freemius.com/mode/dialog/plugin/4553/plan/7330/?trial=paid "The Ethereum ICO Professional plugin")!

https://youtu.be/2lOZw_ov_uY

== FREE Features ==

* To show tokens sell widget insert a simple `[token-ico]` shortcode wherever you like.
* Can be customized with attributes supported: `buybuttontext`, `minimum`, `maximum`, `step`, `placeholder`, `gaslimit`, `tokenname`, `description`
* You also can use fine grained shortcodes for easier customization: `[token-ico-limit label="%s LIMIT!" gaslimit="200001"]`, `[token-ico-input placeholder="Test placeholder"]`, `[token-ico-input-currency showIcons="true" baseCurrency="USD"]`, `[token-ico-buy-button buyButtonText="BUY ME!"]`, `[token-ico-currency-list showIcons="false"  coinList="ETH,BTC"]`
* To show an ICO progress bar widget insert a simple `[token-ico-progress]` shortcode wherever you like. This feature uses https://blockcypher.com `API`. You must use any of the [persistent hash WP plugins](https://codex.wordpress.org/Class_Reference/WP_Object_Cache#Persistent_Cache_Plugins "The persistent hash WP plugins") to overcome its rate limits.
* Use shortcodes to display the current progress value `[token-ico-progress-value]` and percent `[token-ico-progress-percent]`. This feature uses https://blockcypher.com `API`. You must use any of the [persistent hash WP plugins](https://codex.wordpress.org/Class_Reference/WP_Object_Cache#Persistent_Cache_Plugins "The persistent hash WP plugins") to overcome its rate limits.
* To show tokens balance on the current user account use the `[token-ico-balance]` shortcode. If MetaMask is not installed or account is not unlocked, an input field is provided for user account address. `tokenname`, `label` and `placeholder` attributes are supported.
* There is also a shortcode `[token-ico-purchases]` to display a table of recent token purchases by anyone, or by the current user. In the last case if MetaMask is not installed or account is not unlocked, an input field is provided for user account address. `tokenname`, `label` and `placeholder` attributes are supported.
* Use shortcode `[token-ico-referral]` to display a referral address field. User can copy it and send to friends. If they buy tokens while opened this referral link, your `Crowdsale` contract would get a referral address in the `Data` field. Your `Crowdsale` contract should be able to work with it. `url`, `label` and `placeholder` attributes are supported.
* Airdrop is also supported. Just set the minimum allowed setting to zero and the Crowdsale address to your airdrop contract. Note that your airdrop contract should be able to accept zero payments and send some tokens in return.
* This plugin uses Metamask to safely perform the ERC20 token sell operation
* It will show user a link to the Metamask site if the user doesn’t have the Metamask installed
* We use a well known https://etherscan.io API to provide your client an automatic rate calculations to USD and BTC
* Select a list of any currencies supported by the openexchangerates.org to convert the price to.
* You can provide a comma separated list of coins to convert ETH amount inputted by user. This list is shown under the ETH input field.
* Minimum ether amount can be specified to workaround some legal issues.
* Test networks like ropsten or rinkeby are supported. User is warned if he tries to buy tokens from one network, while having MetaMask to point to another network, effectively preventing any losses here.
* This plugin is l10n ready
* Ability to set base currency to display in token sell widget, progress bar and tx table

== PRO Features ==

* You can use a bounty program, if your ERC20 crowdsale contract supports it
* The transaction data to send to your crowdsale contract is supported
* You can use any of the [persistent hash WP plugins](https://codex.wordpress.org/Class_Reference/WP_Object_Cache#Persistent_Cache_Plugins "The persistent hash WP plugins") to overcome the etherscan.io and openexchangerates.org API rate limits. We use the cache to limit the API calling rate to a reasonable value
* Coins and token icons display can be switched on
* Maximum ether amount can be specified to workaround some legal issues
* Private sale seed setting to count funds obtained in a non-Crowdsale contract way
* Custom/private blockchain feature: `Ethereum Node JSON-RPC Endpoint` and other related settings to use the Binance Smart Chain (BSC), Polygon (MATIC) or any other EVM compatible blockchain
* The `[token-ico-referral]` shortcode's default argument name produced in a link is `icoreferral`, like in this example - `https://ethereumico.io?icoreferral=0x476Bb28Bc6D0e9De04dB5E19912C392F9a76535d`, but it can be renamed with the [Referral argument name](https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s131) setting.

> See the [Free vs PRO version differences](https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-free-vs-pro-version-differences/) if need more info on this topic.

== Live Demo ==

See the official site for a Live Demo: [https://ethereumico.io/](https://ethereumico.io/knowledge-base/ico-demo-token-ico-wordpress-plugin/ "The TokenICO Wordpress plugin")

> You can accept fiat money or Bitcoin for your ICO tokens with the [Cryptocurrency Product for WooCommerce](https://wordpress.org/plugins/cryptocurrency-product-for-woocommerce/ "Cryptocurrency Product for WooCommerce") plugin.

== Disclaimer ==

**By using this plugin you accept all responsibility for handling the account balances for all your users.**

Under no circumstances is **ethereumico.io** or any of its affiliates responsible for any damages incurred by the use of this plugin.

Every effort has been made to harden the security of this plugin, but its safe operation depends on your site being secure overall. You, the site administrator, must take all necessary precautions to secure your WordPress installation before you connect it to any live wallets.

You are strongly advised to take the following actions (at a minimum):

- [Educate yourself about cold and hot cryptocurrency storage](https://en.bitcoin.it/wiki/Cold_storage)
- Obtain hardware wallet to store your coins, like [Ledger Nano S](https://www.ledgerwallet.com/r/4caf109e65ab?path=/products/ledger-nano-s) or [TREZOR](https://shop.trezor.io?a=ethereumico.io)
- [Educate yourself about hardening WordPress security](https://codex.wordpress.org/Hardening_WordPress)
- [Install a security plugin such as Jetpack](https://jetpack.com/pricing/?aff=9181&cid=886903) or any other security plugin
- **Enable SSL on your site** if you have not already done so.

> By continuing to use the Ethereum ICO Wordpress plugin, you indicate that you have understood and agreed to this disclaimer.

== Screenshots ==

1. This is how the plugin looks like if the Show icons feature is enabled.
2. The `[token-ico-progress]` display
3. The `[token-ico-balance]` display
4. The `[token-ico-purchases]` display
5. The token symbol and GAS settings
6. The base currency settings
7. Token sale widget settings
8. Advanced settings
9. API keys
10. Displayed currencies list and Show icons flag
11. The ICO Token and Crowdsale contract addresses
12. The ICO Crowdsale contract properties
13. The `[token-ico-progress]` shortcode settings
14. The `[token-ico-referral]` display

== Installation ==

Enter your settings in admin pages and place the `[token-ico]`, `[token-ico-progress]` and other shortcodes wherever you need it.

> Read this step by step guide for more information: [Install and Configure](https://ethereumico.io/knowledge-base/ico-website-install-configure/)

= Main shortcodes example =

`
[token-ico-referral url="https://example.com/crowdsale-ico-page" label="Referal link" placeholder="Input your BSC account address"]

[token-ico-balance tokenname="TSX" label="Referal link" placeholder="Input your BSC account address"]

[token-ico]

[token-ico-progress]

[token-ico-purchases tokenname="TSX" label="Referal link" placeholder="Input your BSC account address"]
`

= Fine grained shortcodes example =

`
[token-ico-limit label="%s LIMIT!" gaslimit="200001"]

[token-ico-input placeholder="Test placeholder"]

[token-ico-input-currency showIcons="true" baseCurrency="USD"]

[token-ico-buy-button buyButtonText="BUY ME!"]

[token-ico-currency-list showIcons="false"  coinList="ETH,BTC"]

[token-ico-progress-value]

[token-ico-progress-percent]

`

= Placeholder =

It is a helper string displayed in the Ether input field for your customer to know where to input Ether amount to buy your tokens.

= Infura.io Api Key =

Register for an infura.io API key and put it in admin settings. It is required to interact with Ethereum blockchain. Use this guide to obtain it: [Get infura API Key](https://ethereumico.io/knowledge-base/infura-api-key-guide/)

= Description =

The `Description` text is displayed immediately after the token purchase widget. It is a good place for some warnings or bounty information.

= Transaction data =

It is an advanced feature. It can be required if your Crowdsale contract can not just accept Ether by send, but need some `payable` method to be called. Do not use if unsure.

= The ICO crowdsale contract address =

The ethereum address of your ICO crowdsale contract. It is the address TokenICO plugin sends Ether to.

> You can input a simple Ethereum address here instead of the Crowdsale contract address. In this case Ether would be sent directly to this your address, but note that you’ll need to send tokens to your customers manually then.

= The ICO token decimals number =

The decimals field of your ICO ERC20 token. The typical value is 1000000000000000000.

> Note that it is different from the `decimals` value in your `Token` contract. If your `Token.decimals` is 18, then you neet to input `10^18 = 1000000000000000000` here. If your `Token.decimals` is 0, then you neet to input `10^0 = 1` here.

= Purchase button =

The Purchase button style has a `button` CSS class and is determined by your WP theme chosen.
You can customize it by adding these code to your `Additional CSS` section in the theme customizing:

`
.button.token-ico-bottom-button-two {
    background-color: #ffd600;
    color: #ffffff;
}
.button.token-ico-bottom-button-two:hover {
    background-color: #ffd6ff;
    color: #ffffff;
}
`

= Progressbar =

`
.twbs .progress {
    background-color: #f5f5f5;
    border-radius: 4px;
}
.twbs .progress-bar {
    background-color: #337ab7;
}
.twbs .progress-meter > .meter > .meter-text {
    color: rgb(160, 160, 160);
}
`

Choose your own colors and additional styles if needed.

= Sidebar small buttons issue =

It is known to have short length button and input area when put on a short width sidebar area. You can fix it with this CSS code:

`
@media (min-width: 992px) {
    .token-ico-shortcode .col-md-5 {
        width: 100%!important;
        max-width: 100%!important;
        flex-basis: 100%!important;
    }
}
`

== Testing ==

You can test this plugin in some test network for free.

=== Testing in ropsten ===

* Set the `Blockchain` setting to `ropsten`
* Set `The ICO token address` setting to 0x6Fe928d427b0E339DB6FF1c7a852dc31b651bD3a or an address of your own token
* Set `The ICO crowdsale contract address` setting to 0x773F803b0393DFb7dc77e3f7a012B79CCd8A8aB9 or an address of your Crowsale contract
* Tune other plugin settings if required
* Buy some tokens with Ropsten Ether
* You can "buy" some Ropsten Ether for free using MetaMask
* Check that proper amount of tokens has been sent to your payment address

=== Testing in rinkeby ===

* Set the `Blockchain` setting to `rinkeby`
* Set `The ICO token address` setting to 0x194c35B62fF011507D6aCB55B95Ad010193d303E or an address of your own token
* Set `The ICO crowdsale contract address` setting to 0x669519e1e150dfdfcf0d747d530f2abde2ab3f0e or an address of your Crowsale contract
* Tune other plugin settings if required
* Buy some tokens with Rinkeby Ether
* You can "buy" some Rinkeby Ether for free here: [rinkeby.io](https://www.rinkeby.io/#faucet)
* Check that proper amount of tokens has been sent to your payment address

== Cache ==

Test if caching is enabled. It is important to overcome API rate limits in production.

Refresh your site page twice and check the HTML source produced for `LOG:` records.

For non-PRO plugin version and for PRO version with Cache plugin not configured properly they would looks like this:

`
LOG: etherscan.io rate_data API is called
LOG: etherscan.io rate_data API call result is stored
LOG: openexchangerates.org API is called
LOG: openexchangerates.org API call result is saved
LOG: etherscan.io total_received API is called
LOG: etherscan.io total_received API call result is stored
`

> Note: make sure to enable the `Object Cache` in the `W3 Total Cache` if you use it.

For [PRO](https://ethereumico.io/product/token-ico-wordpress-plugin/) plugin version with `Cache` properly configured you should see no `LOG:` records most of the time.

See the [ICO Launch: Wordpress Cache Plugin](https://ethereumico.io/knowledge-base/ico-launch-wordpress-cache-plugin/) guide to configure the caching plugin.

== l10n ==

This plugin is localization ready.

Languages this plugin is available now:

* English
* Russian(Русский)

Feel free to [translate](https://translate.wordpress.org/projects/wp-plugins/ethereumico) this plugin to your language.

== Changelog ==

= 2.3.9 =

* MATIC icon is added in the PRO version
* BNB icon not shown issue fix

= 2.3.8 =

* BNB icon is added in the PRO version

= 2.3.7 =

* Disabled `Advanced blockchain` settings in Trial mode fix

= 2.3.6 =

* The name of the argument used in referral links can be renamed now

= 2.3.5 =

* Error display fix

= 2.3.4 =

* False network mismatch error fix

= 2.3.3 =

* `url` attribute is added to the `[token-ico-referral]` shortcode.

= 2.3.2 =

* `tokenname`, `label` and `placeholder` attributes are added to better support non-Ethereum networks.

= 2.3.1 =

* Empty `data` field fix

= 2.3.0 =

* [EIP-1559](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md) support

= 2.2.3 =

* Better BIZ plan support

= 2.2.2 =

* `jQuery is not defined` error fix

= 2.2.1 =

* Unneeded output removed

= 2.2.0 =

* Custom/private blockchain feature: `Ethereum Node JSON-RPC Endpoint` and other related settings to use any EVM compatible blockchain
* ICO Progress bar implemented as a react application to overcome etherscan and bscscan API issues

= 2.1.3 =

* freemius affiliation support

= 2.1.2 =

* Correct display for big token balances that do not fit in the js Number size
* Ether decimals in the input field are calculated according to the token rate setting to allow 1 token purchases

= 2.1.1 =

* MetaMask window on each site page issue is fixed.

= 2.1.0 =

* Firefox CORS issue fix
* web3.js version to 1.3 upgrade

= 2.0.4 =

* freemius.com package version update

= 2.0.3 =

* Display the Account field of the `[token-ico-referral]` shortcode even if MM is not responding

= 2.0.2 =

* infura.io, geth, non-logged MetaMask [fix](https://github.com/INFURA/infura/issues/189#issuecomment-535937835)

= 2.0.1 =

* Show disabled premium settings in the FREE plugin

= 2.0.0 =

* [freemius.com](https://freemius.com) Free/Professional plans business model shift


changelog.txt
== Changelog ==

= 1.11.3 =

* openexchangerates.org fix if App Id is not provided

= 1.11.2 =

* Infura.io legacy API authentication [deprecation](https://blog.infura.io/infura-dashboard-transition-update-c670945a922a)

= 1.11.1 =

* [EIP 1102](https://eips.ethereum.org/EIPS/eip-1102) support
* Load plugin scripts only if some shortcode is used on a page

= 1.11.0 =

* New shortcodes are added for easier token sale widget customization: `[token-ico-limit]`, `[token-ico-input]`, `[token-ico-input-currency]`, `[token-ico-buy-button]`, `[token-ico-currency-list]`
* Shortcodes for current progress value `[token-ico-progress-value]` and percent `[token-ico-progress-percent]` are added

= 1.10.3 =

* Fix the purchases table for contracts which do not send tokens immediately

= 1.10.2 =

* Upgrade web3.js to the latest stable release v0.20.6

= 1.10.1 =

* Fix token amount calculation for purchases table if token rate is set incorrectly.

= 1.10.0 =

* Ability to set base currency to display in token sell widget, progress bar and tx table
* Private sale seed setting
* Correct progress bar value in test networks
* Progress bar values rounding fix

= 1.9.0 =

* l10n support is implemented
* Russian language translation is added

= 1.8.0 =

* Font Awesome copy icon for the affiliate link feature

= 1.7.1 =

* PHP 5.6 compatibility fix

= 1.7.0 =

* Use shortcode `[token-ico-referral]` to display a referral address field. User can copy it and send to friends. If they buy tokens while opened this referral link, your `Crowdsale` contract would get a referral address in the `Data` field. Your `Crowdsale` contract should be able to work with it.
* The `GAS limit` option takes effect in the MetaMask now
* The `GAS price` option is added and takes effect in the MetaMask

= 1.6.5 =

* Mobile display fix for themes with no bootstrap included
* Tx URL fix for test networks

= 1.6.4 =

* Namespaced bootstrap4 is used to prevent any theme conflicts: [bootstrap-namespaced](https://github.com/jonasjancarik/bootstrap-namespaced)
* Bootstrap version is updated from v3 to v4

= 1.6.3 =

* Register bootstrap only if it is not registered yet to not break theme using bootstrap4
* Improoved API call results caching 

= 1.6.2 =

* Unnecessary API calls removed. `LOG` records added to help troubleshoot cache issues.

= 1.6.1 =

* Fix AJAX error in some themes

= 1.6.0 =

* Test networks like ropsten or rinkeby support is added

= 1.5.X or less =

* 1.5.1 Compatibility issues fixed for other plugins using web3 object
* 1.5.0 Airdrop support is added. Just set the minimum allowed setting to zero and the Crowdsale address to your airdrop contract.
* 1.4.0 Two new shortcodes are introduced: `[token-ico-balance]` to show tokens balance on the current user account and 
`[token-ico-purchases]` to display a table of recent token purchases by anyone, or by the current user.
No new admin settings are introduced in this version.
* 1.3.3 Improved compatibility with other plugins and themes.
Improved error handling and reporting.
* 1.3.2 An elementor plugin conflict with the web3.js library is fixed
The ICO progress now shortens values with 1.1K and 1.1M instead of 1100ETH and 1100000ETH respectively.
* 1.3.1 A simple ICO progress bar shortcode is introduced: `[token-ico-progress]`.
The `[token-ico]` shortcode markup is rewritten with bootstrap grid styles for best WP themes and mobile devices compatibility.
The button styles now are WP theme related. It uses the standard `button` CSS class. You can customize it within your theme settings.
* 1.2.0 Minimum and maximum ether amount can be specified to workaround some legal issues.
* 1.1.2 Include bootstrap CSS for some classes used.
* 1.1.1 Bug in the free plugin version is fixed.
* 1.1.0 Select a list of any currencies supported by the openexchangerates.org to convert the price to. 
Optionally, show icons for coins, including your own token.
* 1.0.3 Wrong decimals usage fixed
* 1.0.0 The initial release.


settings/ico.php
<?php


add_filter( 'token_ico_settings_tabs', 'Token_ICO_ico_settings_tabs_hook', 30, 1 );
function Token_ICO_ico_settings_tabs_hook( $possible_screens ) {
    $possible_screens['ico'] = esc_html(__( 'ICO', 'token-ico' ));
    return $possible_screens;
}

add_filter( 'token_ico_get_save_options', 'Token_ICO_ico_get_save_options_hook', 30, 2 );
function Token_ICO_ico_get_save_options_hook( $new_options, $current_screen ) {
    if ('ico' !== $current_screen) {
        return $new_options;
    }

    // ICO options screen

    $new_options['softcap'] = (!empty($_POST['Token_ICO_softcap']) && is_numeric($_POST['Token_ICO_softcap']) ) ? floatval(sanitize_text_field($_POST['Token_ICO_softcap'])) : '';
    $new_options['hardcap'] = (!empty($_POST['Token_ICO_hardcap']) && is_numeric($_POST['Token_ICO_hardcap']) ) ? floatval(sanitize_text_field($_POST['Token_ICO_hardcap'])) : '';
    $new_options['private_sale_seed'] = (!empty($_POST['Token_ICO_private_sale_seed']) && is_numeric($_POST['Token_ICO_private_sale_seed']) ) ? floatval(sanitize_text_field($_POST['Token_ICO_private_sale_seed'])) : '';
    $new_options['icostart'] = (!empty($_POST['Token_ICO_icostart']) /* && is_numeric( $_POST['Token_ICO_icostart'] ) */ ) ? sanitize_text_field($_POST['Token_ICO_icostart']) : '';
    $new_options['icoperiod'] = (!empty($_POST['Token_ICO_icoperiod']) && is_numeric($_POST['Token_ICO_icoperiod']) ) ? intval(sanitize_text_field($_POST['Token_ICO_icoperiod'])) : 30;
    $new_options['tokenAddress'] = (!empty($_POST['Token_ICO_tokenAddress']) /* && is_numeric( $_POST['Token_ICO_tokenAddress'] ) */ ) ? sanitize_text_field($_POST['Token_ICO_tokenAddress']) : '';
    $new_options['crowdsaleAddress'] = (!empty($_POST['Token_ICO_crowdsaleAddress']) /* && is_numeric( $_POST['Token_ICO_crowdsaleAddress'] ) */ ) ? sanitize_text_field($_POST['Token_ICO_crowdsaleAddress']) : '';
    // it should be int, but usually, it is a very big int PHP can not handle, like 1000000000000000000
    $new_options['decimals'] = (!empty($_POST['Token_ICO_decimals']) && is_numeric($_POST['Token_ICO_decimals']) ) ? sanitize_text_field($_POST['Token_ICO_decimals']) : '';
    $new_options['tokenRate'] = (!empty($_POST['Token_ICO_tokenRate']) && is_numeric($_POST['Token_ICO_tokenRate']) ) ? doubleval(sanitize_text_field($_POST['Token_ICO_tokenRate'])) : 1;
//        $new_options['contractABI']      = ( ! empty( $_POST['Token_ICO_contractABI'] )      /*&& is_numeric( $_POST['Token_ICO_contractABI'] )*/ )      ? stripcslashes($_POST['Token_ICO_contractABI'])      : '';
    $new_options['bounty'] = (!empty($_POST['Token_ICO_bounty']) /* && is_numeric( $_POST['Token_ICO_bounty'] ) */ ) ? sanitize_text_field($_POST['Token_ICO_bounty']) : '';
    $new_options['txData'] = (!empty($_POST['Token_ICO_txData']) /* && is_numeric( $_POST['Token_ICO_txData'] ) */ ) ? sanitize_text_field($_POST['Token_ICO_txData']) : '';
    $new_options['referralArgumentName'] = (!empty($_POST['Token_ICO_referralArgumentName']) /* && is_numeric( $_POST['Token_ICO_txData'] ) */ ) ? sanitize_text_field($_POST['Token_ICO_referralArgumentName']) : 'icoreferral';

    return $new_options;
}

add_filter( 'token_ico_print_options', 'Token_ICO_ico_print_options_hook', 30, 2 );
function Token_ICO_ico_print_options_hook( $options, $current_screen ) {
    if ('ico' !== $current_screen) {
        return;
    }
?>

    <tr valign="top">
        <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s20" target="_blank"><?php _e("The ICO token address", 'token-ico'); ?></a></th>
        <td><fieldset>
                <label>
                    <input class="text" name="Token_ICO_tokenAddress" id="Token_ICO_tokenAddress" type="text" maxlength="45" placeholder="<?php _e("Put your ICO token address here", 'token-ico'); ?>" value="<?php echo!empty($options['tokenAddress']) ? esc_attr($options['tokenAddress']) : ''; ?>">
                    <p><?php _e('The Network address of your ICO ERC20 token.', 'token-ico') ?></p>
                </label>
            </fieldset></td>
    </tr>

    <tr valign="top">
        <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s21" target="_blank"><?php _e("The ICO crowdsale contract address", 'token-ico'); ?></a></th>
        <td><fieldset>
                <label>
                    <input class="text" name="Token_ICO_crowdsaleAddress" id="Token_ICO_crowdsaleAddress" type="text" maxlength="45" placeholder="<?php _e("Put your ICO crowdsale token address here", 'token-ico'); ?>" value="<?php echo!empty($options['crowdsaleAddress']) ? esc_attr($options['crowdsaleAddress']) : ''; ?>">
                    <p><?php _e('The Network address of your ICO crowdsale contract. You can input a simple Network address here instead of the Crowdsale contract address. In this case Ether would be sent directly to this your address, but note that you’ll need to send tokens to your customers manually then.', 'token-ico') ?></p>
                </label>
            </fieldset></td>
    </tr>

    <tr valign="top">
        <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s23" target="_blank"><?php _e("The ICO token decimals number", 'token-ico'); ?></a></th>
        <td><fieldset>
                <label>
                    <input class="text" name="Token_ICO_decimals" id="Token_ICO_decimals" type="number" min="0" step="100000000000000000" maxlength="60" placeholder="<?php _e("Put the decimals field of your ICO token contract", 'token-ico'); ?>" value="<?php echo!empty($options['decimals']) ? esc_attr($options['decimals']) : '1000000000000000000'; ?>">
                    <p><?php _e('The decimals field of your ICO ERC20 token. The typical value is 1000000000000000000.', 'token-ico') ?></p>
                </label>
            </fieldset></td>
    </tr>

    <tr valign="top">
        <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s24" target="_blank"><?php echo sprintf(__("The ICO token rate, in %s", 'token-ico'), __("ETH", 'token-ico')); ?></a></th>
        <td><fieldset>
                <label>
                    <input class="text" name="Token_ICO_tokenRate" id="Token_ICO_tokenRate" type="text" placeholder="<?php _e("Put the token rate here", 'token-ico'); ?>" value="<?php echo!empty($options['tokenRate']) ? esc_attr($options['tokenRate']) : '1'; ?>">
                    <p><?php echo sprintf(__('The number of tokens per 1 %1$s, i.e. the token to %1$s exchange rate', 'token-ico'), __("ETH", 'token-ico')) ?></p>
                </label>
            </fieldset></td>
    </tr>

    <tr valign="top">
        <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s25" target="_blank"><?php _e("ICO start date", 'token-ico'); ?></a></th>
        <td><fieldset>
                <label>
                    <input class="text" name="Token_ICO_icostart" id="Token_ICO_icostart" type="date" placeholder="<?php _e("Set the ICO date", 'token-ico'); ?>" value="<?php echo!empty($options['icostart']) ? esc_attr($options['icostart']) : ''; ?>">
                    <p><?php _e('The date when your ICO would start from.', 'token-ico') ?></p>
                </label>
            </fieldset></td>
    </tr>

    <tr valign="top">
        <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s22" target="_blank"><?php _e("ICO period in days", 'token-ico'); ?></a></th>
        <td><fieldset>
                <label>
                    <input class="text" name="Token_ICO_icoperiod" id="Token_ICO_icoperiod" type="number" min="0" step="1" maxlength="3" placeholder="30" value="<?php echo!empty($options['icoperiod']) ? esc_attr($options['icoperiod']) : '30'; ?>">
                    <p><?php _e('The number of days your ICO would be opened.', 'token-ico') ?></p>
                </label>
            </fieldset></td>
    </tr>

    <tr valign="top">
        <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s26" target="_blank"><?php echo sprintf(__("ICO soft cap, in %s", 'token-ico'), __("ETH", 'token-ico')); ?></a></th>
        <td><fieldset>
                <label>
                    <input class="text" name="Token_ICO_softcap" id="Token_ICO_softcap" type="text" placeholder="" value="<?php echo!empty($options['softcap']) ? esc_attr($options['softcap']) : ''; ?>">
                    <?php
                        echo sprintf(__('<p>A soft cap is the amount received at which your crowdsale will be considered successful. It is the minimal amount required by your project for success. You are expected to refund all money if this cap would not be reached.</p><p>This feature uses %1$s `API`. Install some of the %2$spersistent cache WP plugins%3$s to overcome the %4$s API limits. In this case the API would be queried only once per 5 minutes.</p>', 'token-ico')
                            , 'https://blockcypher.com'
                            , '<a target="_blank" href="https://codex.wordpress.org/Class_Reference/WP_Object_Cache#Persistent_Cache_Plugins">'
                            , '</a>'
                            , 'blockcypher.com')
                    ?>
                </label>
            </fieldset></td>
    </tr>

    <tr valign="top">
        <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s27" target="_blank"><?php echo sprintf(__("ICO hard cap, in %s", 'token-ico'), __("ETH", 'token-ico')); ?></a></th>
        <td><fieldset>
                <label>
                    <input class="text" name="Token_ICO_hardcap" id="Token_ICO_hardcap" type="text" placeholder="10000" value="<?php echo!empty($options['hardcap']) ? esc_attr($options['hardcap']) : '10000'; ?>">
                    <?php
                        echo sprintf(__('<p>A hard cap is defined as the maximum amount a crowdsale will receive. The crowdsale is expected to stop after this cap is reached.</p><p>This feature uses %1$s `API`. Install some of the %2$spersistent cache WP plugins%3$s to overcome the %4$s API limits. In this case the API would be queried only once per 5 minutes.</p>', 'token-ico')
                            , 'https://blockcypher.com'
                            , '<a target="_blank" href="https://codex.wordpress.org/Class_Reference/WP_Object_Cache#Persistent_Cache_Plugins">'
                            , '</a>'
                            , 'blockcypher.com')
                    ?>
                </label>
            </fieldset></td>
    </tr>
    <tr valign="top">
        <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s28" target="_blank"><?php _e("Private sale seed", 'token-ico'); ?></a></th>
        <td><fieldset>
                <label>
                    <input class="text" name="Token_ICO_private_sale_seed" id="Token_ICO_private_sale_seed" type="text" placeholder="0" value="<?php echo!empty($options['private_sale_seed']) ? esc_attr($options['private_sale_seed']) : '0'; ?>">
                    <p><?php echo sprintf(__('An amount of funds gained in fiat or non-Ether cryptocurrency in the %s.', 'token-ico')
                        , __("Base currency", 'token-ico'))
                    ?></p>
                </label>
            </fieldset></td>
    </tr>
    <tr valign="top">
      <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s12" target="_blank"><?php _e("Transaction data", 'token-ico'); ?></a></th>
      <td><fieldset>
        <label>
          <input
            class="text" name="Token_ICO_txData" id="Token_ICO_txData" type="text" maxlength="1024" placeholder="0x" value="<?php echo!empty($options['txData']) ? esc_attr($options['txData']) : ''; ?>">
          <p><?php _e('Data to be sent in the transaction. It should starts with \'0x\' without quotes.', 'token-ico'); ?></p>
        </label>
      </fieldset></td>
    </tr>

    <tr valign="top">
      <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s13" target="_blank"><?php _e("Bounty", 'token-ico'); ?></a></th>
      <td><fieldset>
        <textarea
          class="large-text" name="Token_ICO_bounty" id="Token_ICO_bounty" type="text" maxlength="1024000" placeholder="<?php _e("Put your bounty JSON code here", 'token-ico'); ?>"><?php echo!empty($options['bounty']) ? esc_textarea($options['bounty']) : ''; ?></textarea>
        <p><?php _e('The optional JSON array of your bounty values. Note that it should be supported in your crowdsale contract. Example: <code>[[7, 40], [7, 30], [7, 20], [7, 10], [7, 5]]</code>. The 7 number is for days count. The 40, 30, 20, 10, 5 are percents of additional tokens to be sent to buyer for free.', 'token-ico'); ?></p>
      </fieldset></td>
    </tr>

    <tr valign="top">
      <th scope="row"><a href="https://ethereumico.io/knowledge-base/token-ico-wordpress-plugin-configuration/#s131" target="_blank"><?php _e("Referral argument name", 'token-ico'); ?></a></th>
      <td><fieldset>
        <label>
          <input
            class="text" name="Token_ICO_referralArgumentName" id="Token_ICO_referralArgumentName" type="text" maxlength="1024" placeholder="0x" value="<?php echo!empty($options['referralArgumentName']) ? esc_attr($options['referralArgumentName']) : 'icoreferral'; ?>">
          <p><?php echo sprintf(__('The name of the argument used in referral links, like icoreferral here: %s.', 'token-ico'), 'https://ethereumico.io?icoreferral=0x476Bb28Bc6D0e9De04dB5E19912C392F9a76535d'); ?></p>
        </label>
      </fieldset></td>
    </tr>

<?php
}
